crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,585 @@
+                    // The hint is not enforced
+                    if map.len() < MIN_HASH_SIZE {
+                        self.convert_to_list();
+                    }
+                    return;
+                }
+                for (k, v) in iter {
+                    self.insert(k, v);
+                }
+            }
+            AutoMap::Map(map) => {
+                map.extend(iter);
+            }
+        }
+    }
+
+    /// see [HashMap::entry](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry)
+    pub fn entry(&mut self, key: K) -> Entry<'_, K, V> {
+        if self.len() >= MAX_LIST_SIZE {
+            self.convert_to_map();
+        } else if self.len() <= MIN_HASH_SIZE {
+            self.convert_to_list();
+        }
+        match self {
+            AutoMap::List(list) => match list.iter().position(|(k, _)| *k == key) {
+                Some(index) => Entry::Occupied(OccupiedEntry::List { list, index }),
+                None => Entry::Vacant(VacantEntry::List { list, key }),
+            },
+            AutoMap::Map(map) => match map.entry(key) {
+                std::collections::hash_map::Entry::Occupied(entry) => {
+                    Entry::Occupied(OccupiedEntry::Map(entry))
+                }
+                std::collections::hash_map::Entry::Vacant(entry) => {
+                    Entry::Vacant(VacantEntry::Map(entry))
+                }
+            },
+        }
+    }
+}
+
+impl<K: Eq + Hash, V, H: BuildHasher> AutoMap<K, V, H> {
+    /// see [HashMap::get](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get)
+    pub fn get<Q>(&self, key: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        match self {
+            AutoMap::List(list) => {
+                list.iter()
+                    .find_map(|(k, v)| if *k.borrow() == *key { Some(v) } else { None })
+            }
+            AutoMap::Map(map) => map.get(key),
+        }
+    }
+
+    /// see [HashMap::get_mut](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get_mut)
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+        match self {
+            AutoMap::List(list) => {
+                list.iter_mut()
+                    .find_map(|(k, v)| if *k == *key { Some(v) } else { None })
+            }
+            AutoMap::Map(map) => map.get_mut(key),
+        }
+    }
+
+    /// see [HashMap::contains_key](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.contains_key)
+    pub fn contains_key(&self, key: &K) -> bool {
+        match self {
+            AutoMap::List(list) => list.iter().any(|(k, _)| *k == *key),
+            AutoMap::Map(map) => map.contains_key(key),
+        }
+    }
+}
+
+impl<K, V, H> AutoMap<K, V, H> {
+    /// see [HashMap::iter](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter)
+    pub fn iter(&self) -> Iter<'_, K, V> {
```

Comment:
Ditto deduplicating with the `IntoIterator` impl.

Reply:
HashMap does the same: https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter

The code is not duplicated as IntoIterator calls `iter()`